main/modes/system/intro/introMode.c

```diff
@@ -0,0 +1,708 @@
+        detail = iv->tut.tempMessage;
+    }
+
+    int16_t titleY    = 20;
+    iv->bgMenu->title = title;
+    drawMenuMania(iv->bgMenu, iv->renderer, elapsedUs);
+
+    int16_t detailYmin = titleY + iv->bigFont.height + 1 + 10;
+    int16_t detailYmax = TFT_HEIGHT - 20;
+    int16_t detailX    = 10;
+    uint16_t detailH   = textWordWrapHeight(&iv->smallFont, detail, TFT_WIDTH - 20, detailYmax - detailYmin);
+    int16_t detailY    = detailYmax - detailH;
+
+    int16_t viewX = (TFT_WIDTH - iv->swadgeViewWidth) / 2;
+    introDrawSwadge(elapsedUs, viewX, titleY + iv->bigFont.height + 1 + 25, iv->buttons, iv->joysticks);
+
+    const char* remaining
+        = drawTextWordWrap(&iv->smallFont, c000, detail, &detailX, &detailY, TFT_WIDTH - 5, detailYmax);
+    if (NULL != remaining)
+    {
+        ESP_LOGI("Intro", "Remaining text: ...%s", remaining);
+        // TODO handle remaining text sensibly
+    }
+}
+
+/**
+ * This function is called when the display driver wishes to update a
+ * section of the display.
+ *
+ * @param disp The display to draw to
+ * @param x the x coordiante that should be updated
+ * @param y the x coordiante that should be updated
+ * @param w the width of the rectangle to be updated
+ * @param h the height of the rectangle to be updated
+ * @param up update number
+ * @param numUp update number denominator
+ */
+static void introBackgroundDrawCallback(int16_t x, int16_t y, int16_t w, int16_t h, int16_t up, int16_t upNum)
+{
+    fillDisplayArea(x, y, x + w, y + h, c555);
+}
+
+static void introDacCallback(uint8_t* samples, int16_t len)
+{
+#define SPK_SILENCE (INT8_MIN)
+    if (iv->playingSound)
+    {
+        if (iv->soundProgress + len >= iv->soundSize)
+        {
+            size_t send = (iv->soundSize - iv->soundProgress);
+            memcpy(samples, iv->sound + iv->soundProgress, send);
+            memset(samples + send, SPK_SILENCE, len - send);
+            iv->playingSound = false;
+        }
+        else
+        {
+            memcpy(samples, iv->sound + iv->soundProgress, len);
+            iv->soundProgress += len;
+        }
+    }
+    else
+    {
+        memset(samples, SPK_SILENCE, len);
+    }
+}
+
+static void introTutorialCb(const tutorialState_t* state, const tutorialStep_t* prev, const tutorialStep_t* next,
+                            bool backtrack)
+{
+    ESP_LOGI("Intro", "'%s' Triggered!", prev->title);
+
+    if (next == (buttonsSteps + 7))
+    {
+        ESP_LOGI("Intro", "Oh it's the one we want: %s", next->title);
+        iv->quickSettingsOpened = true;
+        openQuickSettings();
+    }
+    else if (next == (buttonsSteps + 8))
+    {
+        iv->quickSettingsOpened = false;
+    }
+    else if (next != NULL && next->trigger.type == NO_TRIGGER)
+    {
+        setTutorialCompletedSetting(true);
+        switchToSwadgeMode(&mainMenuMode);
+        ESP_LOGI("Intro", "Last trigger entered!");
+
+        /*if (iv->curSection < (introSections + ARRAY_SIZE(introSections)))
```

Comment:
Ah, this'd be why the touchpad intro didn't happen

Reply:
Hmm, I forgot this was in there. I think the touchpad parts will need a lot more work so I'll leave it with just the buttons for now. I think I'll also need to add support for a slight delay between finishing a step and moving to the next one so transitions are a bit less jarring so that will help with this.